rasa/core/policies/flow_policy.py

```diff
@@ -695,17 +695,31 @@ def _run_step(
                     and previous_flow_step
                     and current_frame.frame_type == StackFrameType.CORRECTION
                 ):
-                    # TODO: we need to figure out how to actually
-                    #    "undo" the changed slots
                     corrected_slots = tracker.get_slot(CORRECTED_SLOTS_SLOT)
                     if corrected_slots:
-                        self._correct_flow_position(
-                            corrected_slots, previous_flow_step, previous_flow, tracker
-                        )
-                    else:
-                        # TODO: we need to figure out how to actually "undo" the
-                        #    changed slots
-                        pass
+                        if (
+                            tracker.latest_action_name
+                            == "utter_corrected_previous_input"
+                        ):
+                            self._correct_flow_position(
+                                corrected_slots,
+                                previous_flow_step,
+                                previous_flow,
+                                tracker,
+                            )
+                        else:
+                            found_first_event = False
+                            for event in reversed(tracker.events):
```

Comment:
What if there are multiple corrected slots?

Can you describe in general the approach you are trying to aim for? Not 100% sure top of my head what the best solution is to identify the previous set value. Is it processing the events in reverse and picking the second to last slotset for each corrected slot? 🤔